src/main/kotlin/com/exactpro/th2/check1/rule/check/CheckRuleTask.kt

```diff
@@ -35,53 +35,66 @@ import java.time.Instant
 /**
  * This rule checks for the presence of a single message in the messages stream.
  */
+
 class CheckRuleTask(
     ruleConfiguration: RuleConfiguration,
     startTime: Instant,
     sessionKey: SessionKey,
-    private val protoMessageFilter: RootMessageFilter,
+    protoMessageFilter: RootMessageFilter,
     parentEventID: EventID,
     messageStream: Observable<StreamContainer>,
     eventBatchRouter: MessageRouter<EventBatch>
 ) : AbstractCheckTask(ruleConfiguration, startTime, sessionKey, parentEventID, messageStream, eventBatchRouter) {
 
-    private val messageFilter: SailfishFilter = SailfishFilter(
-        CONVERTER.fromProtoPreFilter(protoMessageFilter),
-        protoMessageFilter.toCompareSettings()
-    )
-    private val metadataFilter: SailfishFilter? = protoMessageFilter.metadataFilterOrNull()?.let {
-        SailfishFilter(
-            CONVERTER.fromMetadataFilter(it, METADATA_MESSAGE_NAME),
-            it.toComparisonSettings()
+    protected class Refs(
+        rootEvent: Event,
+        val protoMessageFilter: RootMessageFilter,
+        val messageFilter: SailfishFilter,
+        val metadataFilter: SailfishFilter?
+    ) : AbstractCheckTask.Refs(rootEvent)
+
+    override val refsKeeper = RefsKeeper(
+        Refs(
+            rootEvent = createRootEvent(),
```

Comment:
Why do we create root event in each rule?

Reply:
Because we need to pass it to Refs constructor which is called in each rule class. And we don't using inner classes (in this case we could create root event inside AbstractCheckTask.Refs class).